modal/_partial_function.py

```diff
@@ -922,3 +922,174 @@ def wrapper(
+            original_func = obj.raw_f
+        else:
+            original_func = obj
+
+        @functools.wraps(original_func)
+        def torchrun_wrapper(*args, **kwargs):
+            # Check if we're in data mode (torchrun has set these env vars)
+            if all(var in os.environ for var in ["RANK", "LOCAL_RANK", "WORLD_SIZE"]):
+                # We're in data mode - run the original function
+                return original_func(*args, **kwargs)
+            else:
+                # We're in orchestration mode
+                import sys
+                import subprocess
+                import tempfile
+                import json
+                import modal
+
+                # Get cluster info
+                cluster_info = modal.experimental.get_cluster_info()
+                container_rank = cluster_info.rank
+                main_ip_addr = cluster_info.container_ips[0]
+                n_nodes = len(cluster_info.container_ips)
+
+                # Create a Python script that torchrun will execute
+                # This script needs to reconstruct the function call
+                script_template = """
+import sys
+import os
+import json
+
+# Set up the environment to find the user's code
+sys.path.insert(0, '{working_dir}')
+
+# Import the function
+from {module_name} import {func_name}
+
+# Load arguments if provided
+args_file = sys.argv[1] if len(sys.argv) > 1 else None
+if args_file and os.path.exists(args_file):
+    with open(args_file, 'r') as f:
+        call_info = json.load(f)
+        args = call_info.get('args', [])
+        kwargs = call_info.get('kwargs', {{}})
+else:
+    args = []
+    kwargs = {{}}
+
+# Call the function
+{func_name}.local(*args, **kwargs)
+"""
+
+                # Get function details
+                module_name = original_func.__module__
+                func_name = original_func.__name__
+                working_dir = os.getcwd()
+
+                # Format the script
+                script_content = script_template.format(
+                    working_dir=working_dir, module_name=module_name, func_name=func_name
+                )
+
+                # Write the script to a temporary file
+                with tempfile.NamedTemporaryFile(mode="w", suffix="_torchrun_wrapper.py", delete=False) as f:
+                    f.write(script_content)
+                    script_path = f.name
+
+                # Save arguments to a temporary file (if any)
+                args_file = None
+                if args or kwargs:
+                    # Convert args/kwargs to JSON-serializable format
+                    # Note: This is a simplification and may not work for all argument types
+                    with tempfile.NamedTemporaryFile(mode="w", suffix="_args.json", delete=False) as f:
+                        json.dump({"args": list(args), "kwargs": kwargs}, f)
+                        args_file = f.name
```

Comment:
j/w but do we need to do this thing where we pass the extra args / kwargs into the subprocess via the filesystem? Could we not pass them through in the cli args?

Reply:
Also if you `modal run` against a function using this decorator, do we still successfully propagate its parameters out to the `modal run` CLI? That would be neat!